type-metadata/src/lib.rs

```diff
@@ -1,7 +1,319 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TypeIdent {
+	pub namespace: Vec<String>,
+	pub ident: IdentKind,
+	pub args: Vec<TypeIdent>,
+}
+
+impl TypeIdent {
+	pub fn new(ident: IdentKind) -> TypeIdent {
+		TypeIdent {
+			namespace: vec![],
+			ident,
+			args: vec![],
+		}
+	}
+	pub fn namespace(mut self, namespace: Vec<String>) -> Self {
+		self.namespace = namespace;
+		self
+	}
+	pub fn args(mut self, args: Vec<TypeIdent>) -> Self {
+		self.args = args;
+		self
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(String),
+
+	// primitives, including build-in and common core/std primitives
+	Bool,
+	Str,
+	Unit,
+	U8,
+	U16,
+	U32,
+	Usize,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	Isize,
+	I64,
+	I128,
+	Array(u16),
+	Vector,
```

Comment:
This should more be a `Slice` and if you mean `std::vec::Vec` by this: Why not just take it as a `Custom` type which it is?
Same for `Option` and `Result`. I don't think we should treat them special just because they are custom types within Rust prelude. Having less enum variants is a win imo.

Reply:
Yep I suggested to use `Custom` for `Option/Result/...` types while working on this enum design, @xlc was thinking that as these types are so common, using enum would be an optimization since we don't have to use `string` to represent them in encoding/decoding.

Reply:
Slice/Vector needs some special handling ok SDK to handle length prefix and it is such common type in almost every languages. 
While Option can be a custom type, again it is such a common type in almost all the languages so I think it is ok to spacial case it. 
Result is not that common so can go with custom type.

Reply:
I proposed to rename `Vec` to `Slice` since this encoding represents a slice (`len` + `data`) encoding more than a `Vec` that also technically has a capacity.
Also `Slice` is a fundamental Rust type so it is okay for it to be a special case such as `Tuple` and `Array` are.
For the `Option` and `Result` type I would really discourage special casing them for the sake of reducing the complexity (arity) of the kinds structure.

Reply:
Updated: `Option` and `Result` removed from `IdentKind`; `Vector` renamed to `Slice`.